internal/controllers/gateway/gateway_controller.go

```diff
@@ -0,0 +1,581 @@
+		return ctrl.Result{}, nil
+	}
+
+	// if there's any deletion timestamp on the object, we can simply ignore it. At this point
+	// with unmanaged mode being the only option supported there are no finalizers and
+	// the object should be cleaned up by GC promptly.
+	debug(log, gateway, "checking deletion timestamp")
+	if gateway.DeletionTimestamp != nil {
+		debug(log, gateway, "gateway is being deleted, ignoring")
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	// reconciliation assumes unmanaged mode, in the future we may have a slot here for
+	// other gateway management modes.
+	return r.reconcileUnmanagedGateway(ctx, log, gateway)
+}
+
+// reconcileUnmanagedGateway reconciles a Gateway that is configured for unmanaged mode,
+// this mode will extract the Addresses and Listeners for the Gateway from the Kubernetes Service
+// used for the Kong Gateway in the pre-existing deployment.
+func (r *GatewayReconciler) reconcileUnmanagedGateway(ctx context.Context, log logr.Logger, gateway *gatewayv1alpha2.Gateway) (ctrl.Result, error) {
+	// currently this controller supports only unmanaged gateway mode, we need to verify
+	// any Gateway object that comes to us is configured appropriately, and if not reject it
+	// with a clear status condition and message.
+	debug(log, gateway, "validating management mode for gateway") // this will also be done by the validating webhook, this is a fallback
+	existingGatewayEnabled, ok := gateway.GetAnnotations()[UnmanagedAnnotation]
+	if !ok || existingGatewayEnabled == "" {
+		log.Error(ManagedGatewaysUnsupported, fmt.Sprintf("%s=namespace/name annotation is currently required, will not retry", UnmanagedAnnotation), "namespace", gateway.Namespace, "name", gateway.Name)
+		gateway.Status.Conditions = append(gateway.Status.Conditions, metav1.Condition{
+			Type:               string(gatewayv1alpha2.GatewayConditionScheduled),
+			Status:             metav1.ConditionFalse,
+			ObservedGeneration: gateway.Generation,
+			LastTransitionTime: metav1.Now(),
+			Reason:             string(gatewayv1alpha2.GatewayReasonNoResources),
+			Message:            "managed gateways are not currently suppported, this object can not be scheduled",
+		})
+		return ctrl.Result{}, r.Status().Update(ctx, pruneGatewayStatusConds(gateway))
+	}
+
+	// allow for Gateway resources to be configured with "true" in place of the publish service
+	// reference as a placeholder to automatically populate the annotation with the namespace/name
+	// that was provided to the controller manager via --publish-service.
+	debug(log, gateway, "initializing admin service annotation if unset")
+	if existingGatewayEnabled == "true" { // true is a placeholder which triggers auto-initialization of the ref
+		debug(log, gateway, fmt.Sprintf("a placeholder value was provided for %s, adding the default service ref %s", UnmanagedAnnotation, r.PublishService))
+		gateway.Annotations[UnmanagedAnnotation] = r.PublishService
```

Comment:
Can you please point me to an example setter method, I appear to be missing them.

Reply:
There are none FWIW. It's the first time we're implementing a setter method, and I thought it makes sense for one to live there.

Reply:
Ok, then it seems to me that the thing to do for consistency would be just to set the annotation value as normal?

Reply:
What KIC does before this change is that it uses wrappers from `pkackage annotations` to access annotations instead of manipulating them directly.

By not having a setter function we'd have an inconsistent abstraction: reads happening indirectly, but writes happening directly.

Does this rationale make sense to you?

Reply:
Maybe but it just seems like I'm going to start a new trend by doing that here and I don't see the value.

Reply:
The existing read abstraction is kinda pointless. AFAICT we only have those functions because annotation values are always strings, and the Kong route/service/etc. we stuff them into use their actual type. For example, the getter for a bool Kong entity field converts `"true"` to `true`, a list field converts `"foo,bar"` to `[]string{"foo", "bar"}`, etc.

We never do the reverse and stuff Kong values back into an annotation, and don't usually set annotations at all outside of tests (this is, in fact, the only location we do it per `grep -r "\.Annotations\[" | grep -vi test`).

The getters are of fairly limited use themselves. We only use each once within the parser (`grep -o -r -P "annotations.Has.*\(.*\)" ./ | grep -o -P "annotations.Has.*\(.*\)" | sort | uniq -`) so they could be reasonably inlined. I'd say do that instead if you want consistency (every setter other than UnmanagedAnnotation's would be dead code as we have no use for them), but IMO not worth it either way.

Reply:
This comment is holding up the PR merge, since the original review was non-blocking I'm assuming it's OK to move forward for now. If you would like us to follow-up on this let me know and I'll be happy to.